### ISSUE TITLE
HVOX is bundled with Karabo

### ISSUE THREAD

**Issue by deiruch** [i4Ds/Karabo-Pipeline#426]
https://github.com/matthieumeo/hvox should be integrated.

**Comment by Lukas113** [comment]
[hvox](https://github.com/matthieumeo/hvox) has the following dependencies: 
- numpy
- astropy
- matplotlib
- rascil (prob 1.0.0)
- ska-sdp-datamodels
- ska-sdp-func-python
- healpy
------deps until here are already covered by Karabo------
- [ducc0](https://anaconda.org/conda-forge/ducc0) (conda-build)
- [pycsou](https://github.com/matthieumeo/pycsou) (github-repo)


As far as I've seen, [pycsou](https://github.com/matthieumeo/pycsou) has the following dependencies:
- numpy
- scipy
- dask
- matplotlib
- pandas
- joblib
- scikit-image
- numba
------deps until here are already covered by Karabo------
- [pylops](https://anaconda.org/conda-forge/pylops) (conda build)

**Comment by Lukas113** [comment]
The missing conda builds are provided in the [Karabo feedstock] (https://github.com/i4Ds/Karabo-Feedstock/tree/main/.github/workflows).

**Comment by Lukas113** [comment]
I don't think we can integrate `hvox` into Karabo soon. Accidentally I've looked at the main-branch of `pycsou` (even though I've just described in the post above that they use the last commit of the `v2-dev` branch). There are a lot of changes there. They've a lot of dependencies in their [setup.cfg](https://github.com/matthieumeo/pycsou/blob/v2-dev/setup.cfg) including gpu-dependencies. Here are their current dependencies on their setup.cfg:
```
[options]
package_dir =
	= src
packages =
	pycsou
	pycsou_tests
zip_safe = True
include_package_data = True
python_requires = >= 3.9, < 3.12
install_requires =
	numpy >= 1.23, < 2
	scipy >= 1.9, < 2
	dask[array,diagnostics,distributed] >= 2023.3, < 2023.12
	sparse >= 0.13, < 1
	numba >= 0.56, < 1

[options.extras_require]
dev =
	graphviz >= 0.20, < 1
	ipython >= 8.4, < 9
	pre-commit >= 2.19, < 3
	pytest >= 7.1, < 8
	sphinx >= 5.1, < 6
	tox >= 3.25, < 4
	matplotlib >= 3.5, < 4
	sphinx-rtd-theme == 1.1.*, < 2
	scikit-image >= 0.20, < 1
_gpu =
	nvidia-pyindex
	cupy-cuda12x
	cutensor-cu12
	nvidia-cudnn-cu12
	nvidia-cusparse-cu12
	nvidia-nccl-cu12
fft =
	pyfftw >= 0.13, < 1
	finufft >= 2.1, < 3
fft_gpu =
	nvidia-cufft-cu12
jax =
	jax[cpu] >= 0.4.8, < 1
torch =
	torch >= 2.0, < 3
```

There are a lot of issues with these dependencies:
- They're too restrictive, meaning we have fixed some dependencies in Karabo like `dask` which is not solvable with the restrictions of `pycsou`. So we have to update the Karabo dependencies. E.g. they also have `numpy>1.23` but we have restricted it to 1.22 (Not sure why, @fschramka should know more).
- They have gpu-dependencies. To me it seems like they're using cuda-toolkit 12.* but karabo has referencens to cuda-toolkit 11.*. I think this is an issue, but I don't have experience in this field. In addition, I can't find their corresponding conda-build. If I look e.g. at the `cupy` conda-build, they've their versioning matching with the cuda-toolkit (see [here](https://github.com/cupy/cupy/)). So just installing the corresponding `cupy` might solve this, but I'm not sure.
- They've `torch` as dependency. We had a talk about `tensorflow` which is a transversal dependency of Karabo from `tools21cm`. Is it a good idea to include it? If yes, which configuration? Not sure about that.

So to conclude, there are a lot of issues to solve and test before we can add `hvox` to Karabo.

**Comment by deiruch** [comment]
Thanks for the helpful overview. To address at leadt some questions:
- It's ok and expected to update our dependencies to newer versions (e.g. numpy)
- It's also expected that we'll bundle Tensorflow, Torch, JAX, Keras and others eventually.
- GPU dependencies are more dicey. Perhaps we can start with CPU-backends and figure out how to do GPU-accelerated processing later?

**Comment by Lukas113** [comment]
I just remember something that update to `numpy` > 1.22 is a problem, not sure why anymore. As mentioned I think @fschramka knows more.

I can have a look at how we can support GPU-accelerated processing through Karabo. Not yet sure how to approach this but I think I'll figure it out somehow.

**Comment by Lukas113** [comment]
Delayed until further notice. Reasons:
- #424 has a higher priority
- Have to deal with other dependencies first because of reasons described above.
